### DISCUSSION TITLE
Why does `trait FromRequestParts` require `&mut Parts` and not just `&Parts`?

### DISCUSSION THREAD

**Discussion by Logarithmus** [tokio-rs/axum#2190] [Q&A]
Hello. Are there any reasons, apart from historical ones, why does it require a mutable reference? Because it seems semantically wrong to me. And standard extractors don't mutate `Parts`, as far as I know. Yes, in the past `HeaderMap` extractor did mutate in order to take out the headers instead of cloning them. But some time ago `axum` devs changed this because users weren't satisfied with the errors they were getting in some use cases. And cloning headers isn't really as expensive, compared to cloning the body. So in e. g. axum 0.7 maybe it makes sense to replace `&mut` with `&`? For example, both `actix-web` and `rocket` ask for just immutable references: https://docs.rs/actix-web/latest/actix_web/trait.FromRequest.html https://api.rocket.rs/v0.4/rocket/request/trait.FromRequest.html

**Comment by jplatte** [comment]
For implementers of the trait, this is not a restriction, it *increases* the set of things you can do in the extractor. Yes, it's not common to mutate the request parts when extracting, but I find it nice that users are given the option to do so if they want. Is this somehow problematic for a use case you have?

**Comment by davidpdrsn** [comment]
axum has always given a mutable reference. In 0.5 and earlier extractors were given a [&mut RequestParts](https://docs.rs/axum/0.5.17/axum/extract/trait.FromRequest.html) which also allow mutation.

There are some legitimate use cases for that such as WebSocketUpgrade removing the [OnUpgrade](https://docs.rs/axum/0.6.20/src/axum/extract/ws.rs.html#430). Owned of that is required to upgrade the connection.
